anaconda_anon_usage/utils.py

```diff
@@ -2,18 +2,50 @@
 import os
 import sys
 from os.path import dirname, exists
+from threading import RLock
 
-DEBUG = bool(os.environ.get("ANACONDA_ANON_USAGE_DEBUG"))
+DPREFIX = os.environ.get("ANACONDA_ANON_USAGE_DEBUG_PREFIX") or ""
+DEBUG = bool(os.environ.get("ANACONDA_ANON_USAGE_DEBUG")) or DPREFIX
+
+
+# While lru_cache is thread safe, it does not prevent two threads
+# from beginning the same computation. This simple cache mechanism
+# uses a lock to ensure that only one thread even attempts.
+CACHE = {}
+LOCK = RLock()
+
+
+def cached(func):
+    def call_if_needed(*args, **kwargs):
```

Comment:
This is the replacement for the lru_cache decorator. The key difference is that it uses the locking to ensure the wrapped function is only called once, even in a multithreaded situation.

Reply:
Personally, I would feel more comfortable not reinventing caching tools, but I get it where you're coming from. I remember reading about it when `lru_cache` was added to the stdlib, about the internal thread safety and that it's left to the user to combine it with a locking mechanism in multithreaded applications.

An alternative could be (untested), which would lock access to the LRU cached function:

```python
import functools
import threading

LOCK = threading.RLock()

def locking_lru_cache(maxsize=128, typed=False):
    def decorator(func):
        wrapped_func = functools.lru_cache(maxsize=maxsize, typed=typed)(func)

        @functools.wraps(wrapped_func)
        def inner_wrapper(*args, **kwargs):
            with LOCK:
                return wrapped_func(*args, **kwargs)
        return inner_wrapper
    return decorator
```